drivers/usb/device/usb_dc_nrfx.c

```diff
@@ -773,59 +766,81 @@ static inline void usbd_work_process_recvreq(struct nrf_usbd_ctx *ctx,
 	k_mutex_unlock(&ctx->drv_lock);
 }
 
-/* Work handler */
-static void usbd_work_handler(struct k_work *item)
+
+static inline void usbd_work_process_ep_events(struct usbd_ep_event *ep_evt)
 {
-	struct nrf_usbd_ctx *ctx;
-	struct usbd_ep_event *ev;
+	struct nrf_usbd_ctx *ctx = get_usbd_ctx();
+	struct nrf_usbd_ep_ctx *ep_ctx = ep_evt->ep;
 
-	ctx = CONTAINER_OF(item, struct nrf_usbd_ctx, usb_work);
+	__ASSERT_NO_MSG(ep_ctx);
 
-	if (ctx->flags) {
-		if (ctx->flags & BIT(NRF_USB_STATE_CHANGE)) {
-			usbd_handle_state_change(ctx);
-			ctx->flags &= ~BIT(NRF_USB_STATE_CHANGE);
-		}
+	switch (ep_evt->evt_type) {
+	case EP_EVT_SETUP_RECV: {
+		usbd_work_process_setup(ep_ctx);
+		break;
+	}
+	case EP_EVT_RECV_REQ: {
+		usbd_work_process_recvreq(ctx, ep_ctx);
+		break;
+	}
+	case EP_EVT_RECV_COMPLETE:
+		ep_ctx->cfg.cb(ep_ctx->cfg.addr,
+			       USB_DC_EP_DATA_OUT);
+		break;
 
-		if (ctx->flags & BIT(NRF_USB_STATUS_CHANGE)) {
-			usbd_handle_status_change(ctx);
-			ctx->flags &= ~BIT(NRF_USB_STATUS_CHANGE);
+	case EP_EVT_WRITE_COMPLETE:
+		if (ep_ctx->cfg.type == USB_DC_EP_CONTROL) {
+			k_mutex_lock(&ctx->drv_lock, K_FOREVER);
```

Comment:
this one is tricky

The drv_lock is taken when data is send (`usb_dc_ep_write`) which can be called from any thread, possibly with priority lower then workqueue thread. It can happen then that workqueue will take over the CPU while lock is held and system will deadlock.

Reply:
Discussed offline. I think this is not a problem (at least deadlock) as worqueue thread will loose the CPU and low priority task will be able to finish.
Please ignore this one.